### ISSUE TITLE
Universal threading API

### ISSUE THREAD

**Issue by andrzejlisek** [emscripten-core/emscripten#20509]
The emscripten provides two different and incompatible threading APIs. https://emscripten.org/docs/api_reference/wasm_workers.html

I experienced, that c++11 std::thread seems be compatible with PThreads, so I can use std::thread instead of pthread with compiling by this command:
`emcc thrtest.cpp -o thrtest.js -s "EXPORTED_RUNTIME_METHODS=['ccall']" -pthread -sPTHREAD_POOL_SIZE=5`

The second API is provided by compiling with this command:
`emcc thrtest.cpp -o thrtest.js -s "EXPORTED_RUNTIME_METHODS=['ccall']" -sWASM_WORKERS`

I wrote simple test code with thread and mutex:
```
#include <cstdio>
#include <emscripten.h>
#include <string>
#include <thread>
#include <mutex>

std::thread * Thr;

std::mutex MTX;
std::mutex MTX0;
std::string Str;

int Ptrx = 1;

bool UseMTX = true;

void ThrTestProc()
{
    if (UseMTX) MTX0.lock();
    for (int I = 0; I < 26; I++)
    {
        for (int I0 = 0; I0 < 1000; I0++)
        {
            Ptrx = Ptrx * 3;
            Ptrx = Ptrx / 2;
        }
        MTX.lock();
        std::string SX = "X";
        SX[0] = ((char)(I + 65));
        Str = Str + SX;
        MTX.unlock();
    }
    if (UseMTX) MTX0.unlock();
}

void ThrTest()
{
    printf("Thread test start\n");
    for (int I = 0; I < 10; I++)
    {
        Str = "";
        std::thread * Thr1 = new std::thread(ThrTestProc);
        std::thread * Thr2 = new std::thread(ThrTestProc);
        std::thread * Thr3 = new std::thread(ThrTestProc);
        std::thread * Thr4 = new std::thread(ThrTestProc);
        Thr1->join();
        Thr2->join();
        Thr3->join();
        Thr4->join();
        printf("%s\n", Str.c_str());
    }
    printf("Thread test stop\n");
}

EMSCRIPTEN_KEEPALIVE
void ThrTest1()
{
    UseMTX = true;
    ThrTest();
}

EMSCRIPTEN_KEEPALIVE
void ThrTest0()
{
    UseMTX = false;
    ThrTest();
}

int main()
{
    printf("Thread test\n");

#ifdef __EMSCRIPTEN_PTHREADS__
    printf("Posix thread\n");
#endif
    
#ifdef __EMSCRIPTEN_WASM_WORKERS__
    printf("Wasm worker\n");
#endif
    
    return 0;
}
```

If I compile as PThread, this code works very well, but if I compile as WASM worker, the thread does not run.

It is possible to enforce std::thread to use the WASM Worker API?

What is the `std::thread.join()` equivalent in WASM Worker API?

**Comment by RReverser** [comment]
> so I can use std::thread instead of pthread

It's not instead of. `std::thread` is a C++ wrapper around pthread (at least on Unix platforms and Emscripten), that's why it works with that options, whereas Wasm Workers is completely separate Emscripten-specific API and not part of the standard C or C++ library.

**Comment by andrzejlisek** [comment]
The second question is: How to wait for running thread like `std::thread.join()` in Wasm worker?

In the Wasm worker, the `std::mutex` will work?

Where are listed and described all Wasm worker functions available in API?

**Comment by RReverser** [comment]
> Where are listed and described all Wasm worker functions available in API?

On the API documentation page: https://emscripten.org/docs/api_reference/wasm_workers.html

> In the Wasm worker, the std::mutex will work?

From the same page:

> Pthreads can use the Wasm Worker synchronization API, but not vice versa
The multithreading synchronization primitives offered in emscripten/wasm_worker.h (emscripten_lock_*, emscripten_semaphore_*, emscripten_condvar_*) can be freely invoked from within pthreads if one so wishes, but Wasm Workers cannot utilize any of the synchronization functionality in the Pthread API (pthread_mutex_*, pthread_cond_, pthread_rwlock_*, etc), since they lack the needed pthread runtime.

---

You might also want to check out the [API Differences](https://emscripten.org/docs/api_reference/wasm_workers.html#api-differences) table which should help with some other questions.

Generally, as mentioned above, Wasm Workers is completely separate API, so standard library APIs like `std::thread`, `std::mutex` etc won't work.

If you're compiling existing C/C++ code, most of the time you'll want to use standard library APIs (pthread and C++ equivalents) and not Wasm Workers.

**Comment by andrzejlisek** [comment]
I still do not know, how to wait until the thread finish in main thread. The sample code:

```
#include <emscripten/wasm_worker.h>
#include <stdio.h>

void run_in_worker()
{
  printf("Thread started\n");

  // Some heavy computation

  printf("Thread finished\n");
}

int main()
{
  printf("Main started\n");
  emscripten_wasm_worker_t worker = emscripten_malloc_wasm_worker(/*stackSize: */1024);
  emscripten_wasm_worker_post_function_v(worker, run_in_worker);

  // How to wait until thread finish for continue the main thread?

  printf("Main finished\n");
}
```

The expected output is this:
```
Main started
Thread started
Thread finished
Main finished
```

Using Pthreads, the same code will looks like:
```
#include <pthread.h>
#include <stdio.h>

void * run_in_worker(void *arg)
{
  printf("Thread started\n");

  // Some heavy computation

  printf("Thread finished\n");
  return 0;
}

int main()
{
  printf("Main started\n");
  pthread_t worker;
  pthread_create(&worker, NULL, run_in_worker, (void *)NULL);

  pthread_join(worker, NULL);

  printf("Main finished\n");
}
```

How to do the same effect using Wasm worker?

**Comment by sbc100** [comment]
From the wasm_workers.h header it looks like `emscripten_terminate_wasm_worker` is the way to dispose of a worker:

```
// Terminates the given Wasm Worker some time after it has finished executing       
// its current, or possibly some subsequent posted functions. Note that this        
// function is not C++ RAII safe, but you must manually coordinate to release       
// any resources from the given Worker that it may have allocated from the heap  
// or may have stored on its TLS slots.  There are no TLS destructors that would 
// execute.                                                                      
// Exists, but is a no-op if not building with Wasm Workers enabled              
// (-sWASM_WORKERS=0)                                                            
void emscripten_terminate_wasm_worker(emscripten_wasm_worker_t id);    
```

It looks like its up to you to use shared memory communications to determine when a good time to call this function would be (i.e. when its completely the last task you need of it).

I would highly recommend using the pthread API rather then wasm workers unless you have some really good reason not to.

**Comment by andrzejlisek** [comment]
> From the wasm_workers.h header it looks like emscripten_terminate_wasm_worker is the way to dispose of a worker:

This function purpose rather seems to "kill" just initialized or running thread. I asked, how to wait until thread finish, not how to kill running thread.

> I would highly recommend using the pthread API rather then wasm workers unless you have some really good reason not to.

Now, I am a very newbie in Emscripten and WebAssembly, but I have a long practice in C++ with threads (POSIX, QThread, std:thread) and in C# with threads. In this discussion, I rather learn, make experiments, watch how is works, I do not ask for "real" application needs now.

I consider your recommendation, that PThreads is better choice for real projects with threads, even in case, where cross-platform portability and "advanced" features are not important for me. Theoretically, in my projects, which I plan to implement, there is not important, which threading API I will choose, because it will use very basic features:
1. Create and run thread.
2. Create, lock and unlock "mutex".
3. Wait until thread procedure ends.
4. More features are not necessary and there is the main reasons, why I am finding equivalents of mentioned features in both APIs.

I browsed the test/demo codes on repository and it seems, that this https://github.com/emscripten-core/emscripten/blob/main/test/wasm_worker/semaphore_waitinf_acquire.c is nearest to my question. If I understand good, this code demonstrates the following process:
1. Print text before running threads.
3. Run several threads.
5. Wait until all threads are finished.
6. Print text after finish last thread.

The only difference is following: In the mentioned demo code, the "control" is within one more thread, not in main/gui/browser thread, but I am trying to implement "control" in "GUI thread" or "browser thread". If I call "control_thread()" from main() directly, I get error "waiting is not allowed on this thread". Why?

**Comment by sbc100** [comment]
> > From the wasm_workers.h header it looks like emscripten_terminate_wasm_worker is the way to dispose of a worker:
> 
> This function purpose rather seems to "kill" just initialized or running thread. I asked, how to wait until thread finish, not how to kill running thread.

There is no API for that in Wasm Workers.  You would need to setup some kind of signal and wait for that, and then kill it using emscripten_terminate_wasm_worker.   Better still, don't kill it it all and re-use it for another task later.

**Comment by andrzejlisek** [comment]
> There is no API for that in Wasm Workers.

Can I conclude, that strictly `std::thread.join()` or `pthread_join` equivalent not exists in Wasm Workers?
Can I conclude, that waiting until other thread in main thread is not possible in WW and the Pthreads is the only choice, where such implementation is possible?

**Comment by sbc100** [comment]
Yes, there is no equivalent of std::thread.join() or pthread_join with WW... you would need to build such a thing yourself if you needed it.

Blocking the main thread is generally not good idea with either WW or Pthreads.  Waiting synchronously anther thread to do anything at all (not only exit) is a little risky because the page will be inactive for that time and if you block long enough the browser will assume the page has crashed.   So in terms of blocking the main thread I don't see a big difference between pthreads and WW.  The limitations are basically the same.
